### ISSUE TITLE
Database overload fix

### ISSUE THREAD

**Issue by VasiliyNosov** [pulp/pulp_deb#560]
On Associating Content stage pulp worker overloads the database.
It send to db many heavy sql queries (Maybe for each content entity). For example:

```
pulp=# EXPLAIN ANALYZE SELECT "core_content"."pulp_id"
pulp-# FROM "core_content"
pulp-# WHERE ("core_content"."pulp_id" IN
pulp(#        ('07ca0fc6-7b5a-4ef1-a0c0-c2cddb1a744b'::uuid, '5259d2c2-196a-4f9a-94ca-384a6cc0354a'::uuid) AND
pulp(#        NOT ("core_content"."pulp_id" IN (SELECT V0."pulp_id"
pulp(#                                          FROM "core_content" V0
pulp(#                                                   INNER JOIN "core_repositorycontent" V1 ON (V0."pulp_id" = V1."content_id")
pulp(#                                          WHERE V1."pulp_id" IN (SELECT U0."pulp_id"
pulp(#                                                                 FROM "core_repositorycontent" U0
pulp(#                                                                          INNER JOIN "core_repositoryversion" U2 ON (U0."version_added_id" = U2."pulp_id")
pulp(#                                                                          LEFT OUTER JOIN "core_repositoryversion" U3 ON (U0."version_removed_id" = U3."pulp_id")
pulp(#                                                                 WHERE (U0."repository_id" =
pulp(#                                                                        '453dcd4c-733d-4af0-888d-68f4f3a68356'::uuid AND
pulp(#                                                                        U2."number" <= 1 AND
pulp(#                                                                        NOT (U3."number" <= 1 AND U3."number" IS NOT NULL))))));
                                                                                              QUERY PLAN                                                                                              
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Index Only Scan using core_content_pkey on core_content  (cost=15062.24..15074.71 rows=1 width=16) (actual time=220.331..220.339 rows=0 loops=1)
   Index Cond: (pulp_id = ANY ('{07ca0fc6-7b5a-4ef1-a0c0-c2cddb1a744b,5259d2c2-196a-4f9a-94ca-384a6cc0354a}'::uuid[]))
   Filter: (NOT (hashed SubPlan 1))
   Rows Removed by Filter: 2
   Heap Fetches: 0
   SubPlan 1
     ->  Nested Loop  (cost=10765.29..15059.18 rows=1053 width=16) (actual time=20.395..211.267 rows=23267 loops=1)
           ->  Nested Loop  (cost=10764.86..14444.69 rows=1053 width=16) (actual time=20.373..113.451 rows=23267 loops=1)
                 ->  HashAggregate  (cost=10764.44..10774.97 rows=1053 width=16) (actual time=19.959..25.910 rows=23267 loops=1)
                       Group Key: u0.pulp_id
                       Batches: 1  Memory Usage: 2833kB
                       ->  Nested Loop  (cost=418.30..10761.80 rows=1053 width=16) (actual time=1.450..14.619 rows=23267 loops=1)
                             ->  Hash Left Join  (cost=418.15..10730.48 rows=1109 width=32) (actual time=1.433..8.825 rows=23267 loops=1)
                                   Hash Cond: (u0.version_removed_id = u3.pulp_id)
                                   Filter: ((u3.number > 1) OR (u3.number IS NULL))
                                   ->  Bitmap Heap Scan on core_repositorycontent u0  (cost=416.25..10670.39 rows=22171 width=48) (actual time=1.384..5.689 rows=23267 loops=1)
                                         Recheck Cond: (repository_id = '453dcd4c-733d-4af0-888d-68f4f3a68356'::uuid)
                                         Heap Blocks: exact=330
                                         ->  Bitmap Index Scan on core_repositorycontent_repository_id_1946e61b  (cost=0.00..410.71 rows=22171 width=0) (actual time=1.331..1.331 rows=23267 loops=1)
                                               Index Cond: (repository_id = '453dcd4c-733d-4af0-888d-68f4f3a68356'::uuid)
                                   ->  Hash  (cost=1.40..1.40 rows=40 width=20) (actual time=0.027..0.028 rows=51 loops=1)
                                         Buckets: 1024  Batches: 1  Memory Usage: 11kB
                                         ->  Seq Scan on core_repositoryversion u3  (cost=0.00..1.40 rows=40 width=20) (actual time=0.008..0.013 rows=51 loops=1)
                             ->  Memoize  (cost=0.15..0.17 rows=1 width=16) (actual time=0.000..0.000 rows=1 loops=23267)
                                   Cache Key: u0.version_added_id
                                   Cache Mode: logical
                                   Hits: 23266  Misses: 1  Evictions: 0  Overflows: 0  Memory Usage: 1kB
                                   ->  Index Scan using core_repositoryversion_pkey on core_repositoryversion u2  (cost=0.14..0.16 rows=1 width=16) (actual time=0.010..0.010 rows=1 loops=1)
                                         Index Cond: (pulp_id = u0.version_added_id)
                                         Filter: (number <= 1)
                 ->  Index Scan using core_repositorycontent_pkey on core_repositorycontent v1  (cost=0.42..3.49 rows=1 width=32) (actual time=0.003..0.003 rows=1 loops=23267)
                       Index Cond: (pulp_id = u0.pulp_id)
           ->  Index Only Scan using core_content_pkey on core_content v0  (cost=0.43..0.58 rows=1 width=16) (actual time=0.004..0.004 rows=1 loops=23267)
                 Index Cond: (pulp_id = v1.content_id)
                 Heap Fetches: 16495
 Planning Time: 1.718 ms
 Execution Time: 220.624 ms
(37 rows)
```
For more details:
https://explain.tensor.ru/archive/explain/06371df71c8e5cca9697f7e28e51c841:0:2022-06-24#parsed 

I executed this query on readonly replica and it took 220ms. It's too long.

First synchronization (parallel) of four repositories took more than a day. This makes the pulp a useless and extremely unreliable tool, given that any network error can crash this process.

I use on separate vm:
2xpulp-api server (4 CPU, 4GB RAM)
2xpulp-content server (4 CPU, 4GB RAM)
2xpulp-workers with two worker service on each (4 CPU, 4GB RAM)
2xpg node with pgbouncer and postgres controlled by patroni (8 CPU, 16GB RAM)

pulpcore - 3.18.3
pulp-deb - 2.18.0

**Comment by quba42** [comment]
You are almost certainly hitting this issue from Pulpcore: https://github.com/pulp/pulpcore/issues/2557

It is fixed with pulpcore 3.18.5. Please check if updating your pulpcore version to 3.18.5 fixes your issue.

**Comment by quba42** [comment]
I am closing this because I believe it was fixed in pulpcore. Please reopen if you are still encountering this issue.
